Added a new example showcasing a recursive parser for parsing advance…

gialmisi opened the pull request:
…d mathematical expressions in infix notation.

The parser supports binary, unary, and variadic operators that can be customized by modifying the class variables. It therefore goes way beyond any existing examples for artithmetic expression parsers. It also showcases how regex can be utilized to improve the performance of a parser written utilzing pyparsing.

I wrote this parser for another project, where it was needed. I could not find a ready-made parser for my purposes in the examples, or the web in general. Therefore, I am making this pull request in hopes it gets accepted and hopefully lessens the burden of some other future user, who might have a similar need for a parser as I did. I left comments in the code that I made for myself when learning the library. I believe they add to the example, and therefore left them there.

ptmcg commented:
Looking over your parser this weekend, thanks for submitting it. In the meantime, please look at the plusminus package (https://github.com/pyparsing/plusminus). This package might do much of what you want, plus add evaluation as well as parsing.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+        "((garlic**3 - 2**Lb(cosmic)) + Ln(x**2 + 1)) / (Sqrt(Square(y) + LogOnePlus(potato + 3.1)))",
+    ]
+
+    for expr in expressions:
```
Instead of doing your own loop (and handling exceptions), try `infix_parser.expn.run_tests(expressions)`

gialmisi replied:
I was not aware of this. Thanks!

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+    ParserElement,
+    Suppress,
+    delimitedList,
+    infixNotation,
```
Thanks for using the new PEP8 names where possible, you missed a couple (`infix_notation`, `DelimitedList`).

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+    delimitedList,
+    infixNotation,
+    one_of,
+    opAssoc,
```
`OpAssoc` is preferred over `opAssoc`

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+        "Sinh": "Sinh",
+        "Tanh": "Tanh",
+        "Rational": "Rational",
+        "-": "Negate",
```
I don't think this will ever evaluate as "Negate", since "-" is a signop.

gialmisi replied:
This is true. This is an artifact from the application I have worked and I extracted this example from.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+        unary_func_names = Regex(unary_pattern)
+
+        # Define operands
+        integer = pyparsing_common.integer.set_parse_action(
```
Adding parse actions to integer and scientific is not necessary, since the expressions in pyparsing.common already have these same parse actions.

gialmisi replied:
This is nice to know, I guess this also improves performance a little bit to not add explicit parse actions? Anyhow, I removed them from the example.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+[['Max', [[['Ln', ['x']], '+', ['Lb', [['Abs', ['y']]]]], ['Ceil', [[['Sqrt', ['garlic']], '*', 3]]], [['potato', '**', 2], '/', 4], [['Abs', ['cosmic']], '+', 10]]]]
+"""
+
+from operator import or_
```
`or_` is not used anywhere

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+
+    # Supported infix unary operators, i.e., 'Cos(90)'. The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    UNARY_OPERATORS: ClassVar[dict] = {
```
Feels a little misleading to refer to UNARY and VARIADIC as "operators". I see these more as "function names", which you later use to compose `unary_call` and `variadic_call`.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+
+        # The infix operations do not need to be in this list because they are handled by infixNotation() above.
+        # If new binary operations are to be added, they must be updated in the infixNotation() call (the list).
+        self.reserved_symbols: set[str] = symbols_unary | symbols_variadic
```
Does not seem to be used anywhere.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+        self.reserved_symbols: set[str] = symbols_unary | symbols_variadic
+
+        # It is assumed that the dicts in the three class variables have unique keys.
+        self.operator_mapping = {
```
Does not seem to be used anywhere.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+
+    def parse(self, str_expr: str) -> list:
+        """Parse a string expression into a list."""
+        return self.expn.parse_string(str_expr, parse_all=True)
```
This does not return a list, but a `pyparsing.ParseResults `object. You can convert it to a list using `.as_list()`.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+        # Complete regex pattern with exclusions and identifier pattern
+        exclude = f"{'|'.join([*symbols_variadic, *symbols_unary])}"
+        pattern = r"(?!\b(" + exclude + r")\b)(\b[a-zA-Z_][a-zA-Z0-9_]*\b)"
+        variable = Regex(pattern)
```
Adding `.set_name("variable")` will make your error messages nicer looking. Likewise for `integer`, `scientific`, and others.

gialmisi replied:
I believe I added them to all relevant patterns.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+
+    # Supported infix binary operators, i.e., '1+1'. The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    BINARY_OPERATORS: ClassVar[dict] = {
```
I like the type annotations, but I don't think the `ClassVar` is that useful here. The primary purpose behind `ClassVar` is to signal type-checkers to flag cases where an instance defines its own instance variable by the same name (see https://docs.python.org/3/library/typing.html#typing.ClassVar).  `dict[str, str]` would probably be more appropriate, or even define your own TypeAlias for `dict[str, str]`, and use it for all these mapping vars.

ptmcg commented on examples/infix_math_parser.py:
```diff
@@ -0,0 +1,190 @@
+        rparen = Suppress(")")
+
+        # Define keywords (Note that binary operators must be defined manually)
+        symbols_variadic = set(InfixExpressionParser.VARIADIC_OPERATORS.keys())
```
Calling keys() is unnecessary here, since iterating over a dict gives an iterator over its keys.

ptmcg reviewed (commented):
Nice example, glad to see others making use of `infix_notation` like this.

gialmisi commented:
> Looking over your parser this weekend, thanks for submitting it. In the meantime, please look at the plusminus package (https://github.com/pyparsing/plusminus). This package might do much of what you want, plus add evaluation as well as parsing.

Thanks, I wish I had found this before, might have been useful!

As for the other comments, I have addressed them and updated the PR. I always appreciate someone going over code I have written and giving feedback, especially with this level of care. Thanks!

The diffs are somehow messed up in the PR. I should check my IDE settings. Sorry for this. I would be happy if this example is included in pyparsing, but if its contribution to the framework is not enough to warrant a pull, that is fine as well.

ptmcg commented:
I was going to suggest that the unary_func wasn't necessary, since the variadic_func will match a single argument. But this does work to do verification of the number of arguments at parse time for specific functions. Thanks for being patient on this PR.
